url7d636e8121_laforge49_JActor2_tgz-pJ8-org_agilewiki_jactor2_core_revisited_EHJ8/src/org/agilewiki/jactor2/core/impl/mtPlant/ReactorPoolThreadManager.java

```diff
@@ -0,0 +1,173 @@
+package org.agilewiki.jactor2.core.impl.mtPlant;
+
+import org.agilewiki.jactor2.core.impl.mtReactors.MigrationException;
+import org.agilewiki.jactor2.core.impl.mtReactors.PoolThreadReactorMtImpl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * The ReactorPoolThreadManager is used to process a queue of Reactor's
+ * that have non-empty inboxes.
+ * ReactorPoolThreadManager is a thread pool, but it has a simplified API and
+ * assumes that the thread pool has a fixed number of threads.
+ * ReactorPoolThreadManager is also responsible for assigning the threadReference
+ * when a Reactor is run.
+ */
+public final class ReactorPoolThreadManager {
+    final Logger logger = LoggerFactory
+            .getLogger(ReactorPoolThreadManager.class);
+
+    /**
+     * The taskRequest semaphore is used to wake up a thread
+     * when there is a Reactor which hasWork.
+     */
+    final private Semaphore taskRequest = new Semaphore(0);
+
+    /**
+     * The reactors queue holds the reactors which have messages to be processed.
+     */
+    final private ConcurrentLinkedQueue<PoolThreadReactorMtImpl> reactors = new ConcurrentLinkedQueue<PoolThreadReactorMtImpl>();
+
+    /**
+     * When closing is true, the threads exit as they finish their current activity.
+     */
+    private boolean closing = false;
+
+    /**
+     * The threadCount is the number of threads in the thread pool.
+     */
+    private int threadCount;
+
+    /**
+     * The threads in the thread pool.
+     */
+    private ReactorPoolThread threads[] = null;
+
+    /**
+     * Create a ReactorPoolThreadManager
+     *
+     * @param _threadCount         The number of threads to be created.
+     * @param _threadFactory       Used to create the threads.
+     * @param _maxThreadMigrations Limits the number of times a thread will follow a message in succession.
+     */
+    public ReactorPoolThreadManager(final int _threadCount,
+                                    final int _maxThreadMigrations,
+                                    final ThreadFactory _threadFactory) {
+        this.threadCount = _threadCount;
+        final Runnable runnable = new Runnable() {
+            @Override
+            public void run() {
+                final ReactorPoolThread currentThread = (ReactorPoolThread) Thread
+                        .currentThread();
+                while (true) {
+                    try {
+                        taskRequest.acquire();
+                        PoolThreadReactorMtImpl reactor = reactors.poll();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *reactor*. Currently, there are execution paths that do not contain closure statements, for example, when an if condition at line 114 evaluates to true.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).